Document how to label doctest blocks

fredrikekre opened the pull request:
Added a note on how to label doctest blocks. I am not sure it is correct though, its based on my own experiments...
From my understanding the labeled blocks does *not* have to be consecutive (eg. it works if I insert a non-labeled block in between two labeled ones), but they have to be in the same file to share scope?

MichaelHatherly commented:
Thanks! I'd say it would probably be preferable to use named doctests as opposed to `@meta` blocks since they're slightly less of a pain to use, so might be worth reordering the two sections perhaps. Also, "variable scope" is really just separate modules, might be worth making that clear in the docs.

> From my understanding the labeled blocks does not have to be consecutive

Yes, they don't need to be consecutive and you can intersperse different named doctests between each other. They are evaluated sequentially from top to bottom of the file though.

> but they have to be in the same file to share scope?

Correct, each file is sandboxed from all others with regard to doctests/at-example/at-repl blocks.

fredrikekre commented:
Okay, I will fix later.

> Also, "variable scope" is really just separate modules, might be worth making that clear in the docs.

So same-named blocks is gathered in the same module I presume?

> Yes, they don't need to be consecutive and you can intersperse different named doctests between each other. They are evaluated sequentially from top to bottom of the file though.

I will add a note about this too.

MichaelHatherly commented:
> So same-named blocks is gathered in the same module I presume?

Yes, they just refer to the same `Module`.

fredrikekre commented:
Restructured. Can we come up with a better example for the `DocTestSetup` example? E.g. one where whats in the `DocTestSetup` is actually not used in a block.

fredrikekre commented on docs/src/man/doctests.md:
```diff
@@ -115,11 +115,53 @@ Doctesting can be disabled by setting the [`makedocs`](@ref) keyword `doctest =
 This should only be done when initially laying out the structure of a package's
 documentation, after which it's encouraged to always run doctests when building docs.
 
+## Preserve definition between blocks
```
Note to self: definition**s**

mortenpi replied:
Perhaps also "preserv**ing**"? Matches with the previous heading.

fredrikekre commented:
Updated a bit. Also moved the section about disabling doctests to the bottom, since it felt a bit squashed in between sections about how to set up doctesting.

mortenpi commented on docs/src/man/doctests.md:
`````diff
@@ -105,59 +102,102 @@ line where checking should stop, i.e.
+and hence share scope. This can be useful if the same definitions are used in more than
+one block, with for example text, or other doctest blocks, in between. Example:
+
+````markdown
`````
As I was reading through the text, the illustrative example felt a bit long, so I personally might prefer having it be a bit more concise and divided up.

I would propose the following alternative: have something along these lines after the first two sentences to illustrate the problem:

````
```jldoctest
julia> foo = 42
42
```

```jldoctest
julia> println(foo)
ERROR: UndefVarError: foo not defined
```
````

And then show how you can solve the issue with labels:
````
```jldoctest mylabel
julia> foo = 42
42
```

```jldoctest mylabel
julia> println(foo)
42
```
````

I would also move all the explanatory text outside of the example.

mortenpi commented on docs/src/man/doctests.md:
````diff
@@ -105,59 +102,102 @@ line where checking should stop, i.e.
+rendered documentation:
 
-    Some text discussing `df`...
+```@meta
````
Could also be a bit more concise maybe. I might keep only one definition and one doctest block.

mortenpi commented on docs/src/man/doctests.md:
```diff
@@ -105,59 +102,102 @@ line where checking should stop, i.e.
+!!! note
+
+    The `DocTestSetup` value is **re-evaluated** at the start of *each* doctest block
+    and no state is shared between any code blocks. The `DocTestSetup = nothing` is not strictly
```
I started thinking that the no-state statement is not entirely accurate if you allow for labelled blocks, e.g. as in the following example:

````
```@meta
DocTestSetup = quote
    if isdefined(:foo)
        foo += 10
    else
        foo = 42
    end
end
```

```jldoctest
julia> foo += 1
43
```

```jldoctest
julia> foo += 1
43
```

```jldoctest label
julia> foo += 1
43
```

```jldoctest label
julia> foo += 1
54
```
````

But I am not sure if it is worth complicating the language here.

fredrikekre commented:
I updated a bit, thanks for the comments @mortenpi. I think this is much more concise and more clear.

fredrikekre commented:
On another note, why would these changes cause this when I build the docs?

```
Documenter: building cross-references.
 !! Reference for 'Setup-Code' could no be found. [src/man/syntax.md]
```

mortenpi commented on docs/src/man/doctests.md:
`````diff
@@ -105,59 +102,101 @@ line where checking should stop, i.e.
 
-    ```jldoctest
-    julia> using DataFrames
+````markdown
`````
Something going on here as well. I guess this bit can be dropped and the two paragraphs should be merged?

fredrikekre replied:
oops

fredrikekre commented on docs/src/man/doctests.md:
`````diff
@@ -105,59 +102,101 @@ line where checking should stop, i.e.
 ````
 
-## Skipping Doctests
+In the doctest block below The variable `foo` will not be defined in the next block:
`````
I was a bit quick here... Meant to remove everything before `"The..."

mortenpi commented:
Finally got around to merging this. Thanks again for the contribution @fredrikekre!
